### PR TITLE
Fix clippy error to add clippy job on Ci

### LINKED ISSUE #231: Cargo clippy and cargo fmt in CI
- [x] use [this config](https://github.com/RoaringBitmap/roaring-rs/blob/master/rustfmt.toml)
- [ ] add `cargo clippy` and `cargo fmt` to [the CI](https://github.com/meilisearch/milli/blob/main/.github/workflows/test.yml)
- [ ] Fix clippy error in tests

### PR DESCRIPTION
**Pull request by unvalley**
## Related PR
This PR is for #673 

## What does this PR do?
- ~~add `Run Clippy` job to CI (rust.yml)~~
- apply `cargo clippy --fix` command
- fix some `cargo clippy` error manually (but warnings still remain on tests)

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

### CODE REVIEW

**Comment by unvalley** [comment]
NOTE:
CI fails on all warnings now. So, we have to fix all clippy output (warning and error) to add the clippy job.
It needs a lot of file changes.

**Comment by curquiza** [comment]
@unvalley 
Sorry, since we merged https://github.com/meilisearch/milli/pull/665 related to your work, you have a lot of git conflict. Can you fix them?
I would keep at least the addition of the CI in your PR, which is really useful for us! Thank you 🙏

**Comment by unvalley** [comment]
@curquiza No problem, I'll fix the conflicts!

**Comment by unvalley** [comment]
@curquiza I've fixed conflicts. The last commit passes cargo clippy (but as I described, there are still warnings). 
Please take a look?

File: .github/workflows/rust.yml
Comment (curquiza) [inline .github/workflows/rust.yml:69]:
```suggestion
```

**Comment by curquiza** [comment]
For me it's ok for the CI side @ManyTheFish (or some else in @meilisearch/core-team) I let you review the code base part before merging

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
@unvalley looks like the CIs are all failing, could you fix this before the core team review your PR?

**Comment by curquiza** [comment]
bors try

**Comment by unvalley** [comment]
I have to fix all warnings to pass the CI. 

```rust.yml
env:
  CARGO_TERM_COLOR: always
  RUSTFLAGS: "-D warnings"
```

The RUSTFLAGS `-D warnings` denies all warnings and this pr has warnings by clippy now.

**Comment by curquiza** [comment]
@unvalley, how big is it. Do you want to provide another PR before merging this one fixing all the conflicts?

**Comment by unvalley** [comment]
There are 24 errors. So, that's not too big.  I work to fix the errors in this PR.

**Comment by unvalley** [comment]
@curquiza 
There are 24 errors. So, that's not too big.  I work to fix the errors in this PR.

**Comment by curquiza** [comment]
@unvalley thank you very much this is really appreciated 🙏

**Comment by unvalley** [comment]
@curquiza Could you try bors? :pray:

**Comment by curquiza** [comment]
bors try

**Comment by unvalley** [comment]
bors try

**Comment by unvalley** [comment]
it might be bors ready.

**Comment by unvalley** [comment]
#668 is may be overlapped changes with this #659 . 
but reducing #659 diff is useful for reviewers imo😺

**Comment by curquiza** [comment]
Ok @unvalley thanks for letting me know!
We will merge #668 first

**Comment by unvalley** [comment]
Thanks for running CI. 
I need to align the clippy outputs on CI and on my machine.

**Comment by curquiza** [comment]
@unvalley https://github.com/meilisearch/milli/pull/668 has been merged! Can you fix the merge conflicts? 😇

**Comment by ehiggs** [comment]
Hey @unvalley sorry for the rebase I've handed you. If you want, I can do the rebase and make a PR to your branch? You'd need to check your fork if/when that happens.

**Comment by unvalley** [comment]
@ehiggs No worries, Thanks for fixing clippy errors!

File: milli/src/lib.rs
Comment (unvalley) [inline milli/src/lib.rs:2]: imo, when we change the code that has clippy error, we should realize that is allowed.

**Comment by unvalley** [comment]
@curquiza Ok, I fixed conflicts. Could you try bors again?

**Comment by curquiza** [comment]
bors try

**Comment by unvalley** [comment]
oops, it conflicts again.

File: milli/src/lib.rs
Comment (curquiza) [reply milli/src/lib.rs:2]: @ManyTheFish will be the best to answer 😇

**Comment by unvalley** [comment]
@curquiza I fixed conflicts again 🙏🏻

**Comment by unvalley** [comment]
okay, the errors relate to facets refactoring. I'll fix that.

**Comment by curquiza** [comment]
bors try

**Comment by unvalley** [comment]
@curquiza I fixed conflicts again😆 (sorry for trying bors many times!)

**Comment by curquiza** [comment]
bors try

**Comment by unvalley** [comment]
fixed rust fmt🔥

**Comment by curquiza** [comment]
Bors try

**Comment by unvalley** [comment]
@curquiza 
I finally removed the clippy job in this pr to receive a code review for the current code change. 
I'll add the job in [this pr](https://github.com/meilisearch/milli/pull/673).

File: filter-parser/src/lib.rs
Comment (irevoire) [inline filter-parser/src/lib.rs:348]:
```suggestion
/// primary        = (WS* "(" WS* expression WS* ")" WS*) | geoRadius | condition | exists | not_exists | to
```

Is there a reason for this change?

File: milli/src/search/facet/filter.rs
Comment (irevoire) [inline milli/src/search/facet/filter.rs:106]:
```suggestion
                    match ors.len() {
                        [] => (),
                        [or] => ands.push(or),
                        ors => ands.push(FilterCondition::Or(ors)),
```

Can't we do something like that instead?

**Review by irevoire — CHANGES_REQUESTED** [review]
Hey @unvalley, to me we're almost finished, there is just two last things that I would like to check but overall I think we can merge

**Comment by irevoire** [comment]
Hey @unvalley, since this is quite a big PR and you had to rebase a lot of changes we wanted to thank you for your time and accept your contribution for the hacktoberfest even though we're probably not going to merge it before the end of the event.
That's why I gave you the hacktoberfest-accepted tag.
And if you would like to receive some swag from Meilisearch, please complete [this form](https://meilisearch.typeform.com/to/idt5qugY).

**Comment by unvalley** [comment]
@irevoire Thanks for letting me know! I completed the form. I'd like to contribute more.

File: filter-parser/src/lib.rs
Comment (unvalley) [reply filter-parser/src/lib.rs]: I guess this change is by `cargo clippy --fix`. but I have no particular opinion on this.
So I accepted your suggestion, and if the change is needed in other PR, I'll describe about this.

**Comment by curquiza** [comment]
And thanks for being such reactive @unvalley! You are more than welcome to contributing again 😄

File: milli/src/search/facet/filter.rs
Comment (unvalley) [reply milli/src/search/facet/filter.rs:106]: I tried to your suggestion in local and executed clippy, but compile errors occurred.

Screen Shot 2022-11-01 at 21 47 22

type mismatch for `ors`.

Comment (Kerollmops) [reply milli/src/search/facet/filter.rs:106]: Indeed, as this is a `Vec` and not a slice it can't work this way but you can probably do that:
```suggestion
                    match ors.len() {
                        0 => (),
                        1 => ands.push(ors.pop().unwrap()),
                        _ => ands.push(FilterCondition::Or(ors)),
```

Comment (unvalley) [reply milli/src/search/facet/filter.rs]: Thanks for suggestion, I follow the code and it passed clippy run.

**Review by irevoire — APPROVED** [review]
Good to me, since the PR is quite large I'll let someone else from the @meilisearch/core-team review + merge

File: milli/src/search/facet/filter.rs
Comment (irevoire) [reply milli/src/search/facet/filter.rs]: Thanks for implementing the change @unvalley!!

**Review by Kerollmops — APPROVED** [review]
There are a lot of changes but it looks good to me too!
Thank you very much ❤️

bors merge

**Pull request merged**
